Add a New Example IS Plugin for LabVIEW (#37)
ni/instrumentstudio-plugins: labview/examples/Game of Life/README.md

Reviewer: What concepts or techniques does this example provide that aren't covered by the base template from the generator? In other words, what is the purpose of this example vs. any other example that we might add in the future.  I'm asking to tease out whether we should consider adding some of these features to the default template and how we would draw the line for adding new examples in the future. Based on those answers, including some information in the readme about what the user should take away from the example might be useful.

Author: > What concepts or techniques does this example provide that aren't covered by the base template from the generator? In other words, what is the purpose of this example vs. any other example that we might add in the future. I'm asking to tease out whether we should consider adding some of these features to the default template and how we would draw the line for adding new examples in the future.

This example demonstrates how to develop a LabVIEW application that simulates Conway's Game of Life theory as an InstrumentStudio plugin using the InstrumentStudio Plugin SDK APIs.

- **Base Template vs. Example Purpose**:
  - The base template showcases the functionality of the IS Plugin SDK APIs and provides a better starting point for the development of IS plugins.
  - The template does not contain any logic to demonstrate the use of these APIs from an application point of view.
  - While this Game of Life example illustrates how users can integrate and utilize these APIs in their LabVIEW applications to develop them as IS plugins.
  - It serves as a reference for users/developers during the development of their applications.

- **Design Pattern**:
  - The example plugin follows the Queued Message Handler design pattern.
  - While, the existing example-plugin (template plugin) follows the Event Handler design pattern, which contains a single Event Handling loop. Since this example plugin requires a responsive UI, the design pattern is extended to follow the Queued Message Handler pattern by adding a Message Handling loop.
  - The current template ensures flexibility by following a basic single loop-based design pattern, allowing users to decide whether to change the design pattern or follow the existing Event Handler pattern based on their use case and application requirements.

Kindly let me know if any additional details are required and if we can remove or retain the example plugin in the repository's `labview/examples/example-plugin` folder.

> Based on those answers, including some information in the README about what the user should take away from the example might be useful.

I will update the README to include information about the key takeaways for the user from the example.

Author: @jasonmreding, as suggested, I have updated the README to clarify the purpose of this example and am awaiting your review to proceed with this PR.

   Instruments\InstrumentStudio\Addons` by default. This action will require administrative
   permission.
   1. You may optionally install the plug-in files into a subdirectory of the Addons directory so
      its contents do not conflict with other plug-ins.
   2. In addition to the default Addons directory, you can also install the plug-in files to a
      user-defined directory configured in the InstrumentStudio preferences dialog for plug-ins.
        - This user-defined directory allows for the installation of plug-ins without requiring
          administrative permissions.
        - To configure this preference, InstrumentStudio must be launched with administrator
          privileges.
5. Now, open InstrumentStudio, create a new InstrumentStudio project, and save it.
6. From the project window, click `Manual Layout` to open the Edit Layout dialog.
   1. Navigate to `Add-Ons` -> `NI Example Plugins` -> `Conway's Game of Life`.
   2. Create a large panel.
7. Once the SFP is open, an instance of the plug-in VI will start running in the backend.
8. By default, the controls in the SFP will load with either their default values or the last saved

>    values. You can adjust the controls to change the game board size and the update interval between
>    generations as needed.

Reviewer 2: ```suggestion
   values. 
   - Adjust the controls to change the game board size and the time interval between number of
   generations on the board.
```

9. Click on the `Start` button to run the simulation to visualise Conway's Game of Life theory on
   the graph/game board.
   1. The simulation will run indefinitely as long as there are 'births' and 'deaths' occurring on
      the graph.

>    2. Click on the `Stop` button to stop the simulation.

Reviewer 2: Another point to mention that Time Interval can be updated during runtime and the board will be updated based on the time interval

10. The plug-in VI instance will stop running on closure of the SFP or closure of the
    InstrumentStudio project window or removal of the SFP from layout.

## Developer Guide

### Make changes to the example plug-in

- The implementation logic for the Game of Life InstrumentStudio plug-in example is found in the
  top-level plug-in VI `Game of Life.vi`.
- Add your logic/edit the existing logic in the top-level VI to make changes to the plug-in.

### How to add new UI elements

- Add controls and indicators to the front panel of the top-level plug-in VI.
- Capture events associated with the controls in the Event Handling Loop and pass the data to
  Message Handling Loop through the `Queue Driver.vi` for further operations.
- Add the necessary logic is respective states of Message handling loop to handle subsequent
  processing and actions.

> Note: This example plug-in follows Queued Message Handler design pattern to ensure a responsive UI,
> while the choice of design pattern can vary based on the application needs.

Reviewer 2: ```suggestion
Note: This example plug-in uses the Queued Message Handler design pattern to handle a responsive UI,
while the choice of design pattern can vary based on the application needs.
```
